src/Streamly/Internal/Data/Array/Stream/Mut/Foreign.hs

```diff
@@ -194,29 +208,114 @@ compact :: (MonadIO m, Storable a)
+    step (Just buf) arr =
+        let len = MArray.byteLength buf + MArray.byteLength arr
+         in if len > n
+            then return $ ParserD.Done 1 buf
+            else do
+                buf1 <-
+                    if MArray.byteCapacity buf < n
+                    then liftIO $ MArray.realloc n buf
+                    else return buf
+                buf2 <- MArray.splice buf1 arr
+                return $ ParserD.Partial 0 (Just buf2)
+
+    extract Nothing =
+        throwM
+            $ ParserD.ParseError $ functionPath ++ ": The slice buffer is empty"
+    extract (Just buf) = return buf
+
+    functionPath =
+        "Streamly.Internal.Data.Array.Stream.Mut.Foreign.compactLEParserD"
+
+-- | Coalesce adjacent arrays in incoming stream to form bigger arrays of a
+-- maximum specified size. When we coalesce multiple arrays if the size would
+-- exceed the specified size we stop the fold. It might be possible to get an
+-- array that is smaller than the specified size if all the arrays together in
+-- the incoming stream are smaller than the specified size.
+--
+-- /Unimplemented/
+{-# INLINE_NORMAL compactGEFold #-}
+compactGEFold ::
+       (MonadIO m, Storable a)
+    => Int -> FL.Fold m (Array a) (Array a)
+compactGEFold n = Fold step initial extract
+
+    where
+
+    initial =
+        return
+            $ if n <= 0
+              then error
+                       $ functionPath
+                       ++ ": the size of arrays ["
+                       ++ show n ++ "] must be a natural number"
+              else FL.Partial Nothing
+
+    step Nothing arr =
+        return
+            $ let len = MArray.byteLength arr
+               in if len >= n
+                  then FL.Done arr
+                  else FL.Partial (Just arr)
+    step (Just buf) arr = do
+        let len = MArray.byteLength buf + MArray.byteLength arr
+        buf1 <-
+            if MArray.byteCapacity buf < len
+            then liftIO $ MArray.realloc len buf
+            else return buf
+        buf2 <- MArray.splice buf1 arr
+        if len >= n
+        then return $ FL.Done buf2
+        else return $ FL.Partial (Just buf2)
+
+    extract Nothing = error $ functionPath ++ ": The slice buffer is empty"
+    extract (Just buf) = return buf
+
+    functionPath =
+        "Streamly.Internal.Data.Array.Stream.Mut.Foreign.compactGEFold"
 
 -- | Coalesce adjacent arrays in incoming stream to form bigger arrays of a
 -- maximum specified size in bytes.
 --
 -- /Internal/
-compactLE :: (MonadIO m {-, Storable a-}) =>
-    Int -> SerialT m (Array a) -> SerialT m (Array a)
+compactLE ::
+       (MonadIO m, MonadThrow m, Storable a)
+    => Int -> SerialT m (Array a) -> SerialT m (Array a)
 compactLE n (SerialT xs) =
-    SerialT $ D.toStreamK $ D.foldMany (compactLEFold n) (D.fromStreamK xs)
+    SerialT $ D.toStreamK $ D.parseMany (compactLEParserD n) (D.fromStreamK xs)
 
+-- XXX This isn't possible unless we mutate the array of the incoming stream.
```

Comment:
I've had something else in mind! What I meant was this isn't possible idiomatically.
Cant use a `parseMany` or `foldMany` to implement this. This can only be implemented non-idiomatically

Reply:
You can implement it idiomatically if you use ArrayStream folds.